src/runtime/optional.lisp

```diff
@@ -0,0 +1,97 @@
+;;;; optional.lisp
+;;;;
+;;;; This package contains the underlying implementation of the
+;;;; Optional type for Coalton. The idea is to avoid unnecessary
+;;;; boxing by defining a structure that represents None nested in N
+;;;; Somes. Consider the general case of the type (Optional^M :t). In
+;;;; this case, we have exactly M+1 possible cases to match
+;;;; against. We have the cases (Some^N (None)) for N from 0 to M-1
+;;;; and we have the case (Some^M x) where x is of type :t. Therefore,
+;;;; by defining a struct CL-OPTIONAL with one slot, DEPTH, we can
+;;;; always represent (Optional^M :t) as either an element of type :t
+;;;; or as a CL-OPTIONAL with a depth between 0 and M-1.
+;;;;
+;;;; For example, (lisp (Optional (Optional :t)) () 5) is
+;;;; unambiguously equal to (Some (Some 5)).
+
+(defpackage #:coalton-impl/runtime/optional
+  (:use
+   #:cl)
+  (:export
+   #:cl-none
+   #:cl-none-p
+   #:cl-some
+   #:cl-some-p
+   #:unwrap-cl-some))
+
+(in-package #:coalton-impl/runtime/optional)
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (declaim (inline make-cl-optional))
+  (defstruct (cl-optional (:constructor make-cl-optional (depth)))
+    "A CL-OPTIONAL represents of DEPTH N represents the Coalton
+value (Some^N None)."
+    (depth 0 :type (unsigned-byte 16) :read-only t)))
+
+#+sbcl
+(cl:declaim (sb-ext:freeze-type cl-optional))
+
+(defmethod print-object ((opt cl-optional) stream)
```

Comment:
I wouldn't attempt to fix this, but technically, this doesn't respect some aspects of pretty printing, like the print width and the like.

Don't worry about it though.